### PR TITLE
support negative offset features (and add IDA tests)

### LINKED ISSUE #197: support negative offset features
currently we have a lint that rejects rules that contain negative numbers, specifically constants and offsets. i think this is primarily because its can be difficult to determine how immediate operands should be interpreted as signed or unsigned values. therefore, this restriction makes sense for constants/numbers; however, the way that we extract offsets (from memory reference operands) does support negative values. so, should we relax the lint and only enforce it on numbers/constants, but not offsets?

related: https://github.com/fireeye/capa-rules/pull/52

@mike-hunhoff @mr-tz

### LINKED ISSUE #202: add tests for IDA Pro backend
because otherwise i'm not sure if my changes are breaking things!

### PR DESCRIPTION
**Pull request by williballenthin**
codifies our decision to support offset features with negative values #197 by relaxing the lint and updating the IDA extractor.

in order to demonstrate that the IDA changes work, i implemented tests for the IDA extractor. these will not run during CI, but we can invoke them manually during development. they should be run with the `mimikatz.exe_` database open.

closes #197 
closes #202

### CODE REVIEW

File: capa/features/extractors/ida/insn.py [hunk 1]
```diff
@@ -144,12 +144,13 @@ def extract_insn_offset_features(f, bb, insn):
             continue
         p_info = capa.features.extractors.ida.helpers.get_op_phrase_info(op)
         op_off = p_info.get("offset", 0)
-        if 0 == op_off:
-            continue
```
Comment (williballenthin) [inline capa/features/extractors/ida/insn.py:148]: for consistency with the viv extractor, we support offset=0x0.

**Comment by williballenthin** [comment]
image

File: capa/features/extractors/ida/insn.py [hunk 2]
```diff
@@ -144,12 +144,13 @@ def extract_insn_offset_features(f, bb, insn):
...
         if idaapi.is_mapped(op_off):
             # Ignore:
             #   mov esi, dword_1005B148[esi]
             continue
+
+        op_off = capa.features.extractors.helpers.twos_complement(op_off, 32)
```
Comment (mike-hunhoff) [inline capa/features/extractors/ida/insn.py:152]: What about 8 and 16?

File: tests/test_ida_features.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+logger = logging.getLogger("test_ida_features")
+
+
+def check_input_file():
+    import idautils
+
+    wanted = "5f66b82558ca92e54e77f216ef4c066c"
+    # some versions of IDA return a truncated version of the MD5.
+    # https://github.com/idapython/bin/issues/11
+    found = idautils.GetInputFileMD5().rstrip(b"\x00").decode("ascii").lower()
```
Comment (mike-hunhoff) [inline tests/test_ida_features.py:22]: Good to know!

File: tests/test_ida_features.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+    f = idaapi.get_func(0x4011FB)
+    features = extract_function_features(f)
+    assert len(features[capa.features.basicblock.BasicBlock()]) == 15
+
+
+if __name__ == "__main__":
+    print("-" * 80)
+
+    # invoke all functions in this module that start with `parse_`
```
Comment (mike-hunhoff) [inline tests/test_ida_features.py:276]: change `_parse` to `_test`.

File: capa/features/extractors/ida/insn.py (hunk 2 above)
Comment (mr-tz) [reply capa/features/extractors/ida/insn.py:152]: or 64?

Comment (williballenthin) [reply capa/features/extractors/ida/insn.py:152]: i got the impression that IDA returns the offset as twos complement in a u32, never as u8/u16/u64. if this is wrong, i'd very much like to fix it.

according to some sources, like [here](https://stackoverflow.com/q/31853189/87207), there is no 64-bit offset/displacement.

**Pull request merged**